File: backend/app/schemas/query.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, Any


class QueryRequest(BaseModel):
    connection_id: int
    natural_language: str
    auto_execute: bool = False


class SQLGenerateResponse(BaseModel):
    generated_sql: str
    explanation: str
    tables_used: list[str]
    optimized_sql: Optional[str] = None
    optimization_notes: Optional[str] = None


class QueryExecuteRequest(BaseModel):
    connection_id: int
    sql: str
    history_id: Optional[int] = None


class QueryResult(BaseModel):
    columns: list[str]
    rows: list[list[Any]]
    row_count: int
    execution_time_ms: float
    suggested_chart: Optional[str] = None


class QueryHistoryOut(BaseModel):
    id: int
    natural_language: str
    generated_sql: Optional[str]
    explanation: Optional[str]
    optimized_sql: Optional[str]
    execution_time_ms: Optional[float]
    row_count: Optional[int]
    is_successful: bool
    error_message: Optional[str]
    chart_type: Optional[str]
    tables_used: Optional[list]
    created_at: datetime

    model_config = {"from_attributes": True}


class DashboardStats(BaseModel):
    total_queries: int
    successful_queries: int
    failed_queries: int
    avg_execution_time_ms: float
    most_queried_tables: list[dict]
    queries_by_day: list[dict]
    total_connections: int

File: frontend/store/query.ts
import { create } from "zustand";

export interface QueryResult {
  columns: string[];
  rows: unknown[][];
  row_count: number;
  execution_time_ms: number;
  suggested_chart: string | null;
}

export interface GeneratedSQL {
  generated_sql: string;
  explanation: string;
  tables_used: string[];
  optimized_sql: string | null;
  optimization_notes: string | null;
}

interface QueryState {
  selectedConnectionId: number | null;
  naturalLanguage: string;
  generatedSQL: GeneratedSQL | null;
  queryResult: QueryResult | null;
  isGenerating: boolean;
  isExecuting: boolean;
  error: string | null;
  historyId: number | null;

  setConnection: (id: number | null) => void;
  setNaturalLanguage: (nl: string) => void;
  setGeneratedSQL: (sql: GeneratedSQL | null) => void;
  setQueryResult: (result: QueryResult | null) => void;
  setGenerating: (v: boolean) => void;
  setExecuting: (v: boolean) => void;
  setError: (e: string | null) => void;
  setHistoryId: (id: number | null) => void;
  reset: () => void;
}

export const useQueryStore = create<QueryState>((set) => ({
  selectedConnectionId: null,
  naturalLanguage: "",
  generatedSQL: null,
  queryResult: null,
  isGenerating: false,
  isExecuting: false,
  error: null,
  historyId: null,

  setConnection: (id) => set({ selectedConnectionId: id }),
  setNaturalLanguage: (nl) => set({ naturalLanguage: nl }),
  setGeneratedSQL: (sql) => set({ generatedSQL: sql }),
  setQueryResult: (result) => set({ queryResult: result }),
  setGenerating: (v) => set({ isGenerating: v }),
  setExecuting: (v) => set({ isExecuting: v }),
  setError: (e) => set({ error: e }),
  setHistoryId: (id) => set({ historyId: id }),
  reset: () =>
    set({ generatedSQL: null, queryResult: null, error: null, historyId: null }),
}));
